### ISSUE TITLE
support for time coordinates in Common

### ISSUE THREAD

**Issue by pebau** [opengeospatial/ogcapi-common#85]
Like space, time is an intrinsic property of the physical world our geo standards model, across most (if not all) services. Therefore, time handling should be homogenized, ie: reside in Common.

Coverages - out of sheer need - have an elaborate time model treated with exactly the same mechanisms as space: time has a CRS, time can have regular and irregular axes, time is a dimension amenable to subsetting, and even more than one time axis can be modelled intuitively. Therefore, the coverage time model is a prime candidate for use in Common.

**Comment by bradh** [comment]
I think a lowest-common-denominator time concept for Common would be useful, as long it can be extended. 

In general, Common should be the intersection of what all services need though, not the union.

**Comment by dblodgett-usgs** [comment]
@bradh -- is that a position that's been decided on? IMHO, common should contain anything that multiple API specs want to have work in a common way. 

I'm not sure what all is implied by:
> Therefore, the coverage time model is a prime candidate for use in Common.

How is that different from [what's already in OGC-API Features](http://docs.opengeospatial.org/is/17-069r3/17-069r3.html#_parameter_datetime)?

**Comment by pebau** [comment]
@bradh : re "intersection instead of union", time likely is part of the intersection. Fully agreed on having something that does imply things in a simple way, and allows to do advanced stuff without breaking principles or introducing extra tweaks. That's what the coverage community has established, after long discussion, with CIS &WCS back in 2012: space, time, and "any other" axis are treated in exactly the same way. There is just one `subset` parameter for each axis ("dimension") that can be used, left out, and combined / repeated for other axes, ad libitum.

@dblodgett-usgs : some features are mentioned above already, but maybe just look at an example:
- JSON: http://schemas.opengis.net/cis/1.1/json/examples/30_4D_height+time.json
- XML: http://schemas.opengis.net/cis/1.1/gml/examples-1.1/30_4D_height+time.xml
- subsetting (which differentiates into trimming and slicing: ...?SUBSET=Long(100,120) & SUBSET=Lat(50,60) & SUBSET=time("2009-11-06T23:20:52")
- more about CRSs: http://external.opengeospatial.org/twiki_public/CRSdefinitionResolver/WebHome

**Comment by bradh** [comment]
@pebau I think the time model in WCS is probably needed for coverages (e.g. to support weather forecasts, where there are multiple times involved). It adds complexity if everyone needs to understand it to get their maps though, and subsetting may not make sense for a map. 

So I'd like to see time in Common, at the "what makes sense everywhere" level. I'd like to see the advanced part of time put only where it is needed. 

In terms of a philosophy, I'm looking for Common to not become the new GML.

**Comment by pebau** [comment]
guess I agree about GML ;-) ...but temporal subsetting you don't have to apply to a map where time is irrelevant, so there is no burden in cases where you don't need it.  For example, in a snippet like
   ...?SUBSET=Long(100,120) & SUBSET=Lat(50,60)
you would get a 2D result for a map, a 3D result for a timeseries. Conversely,
   ...?SUBSET=time("2009-11-06T23:20:52")
can return a 2D map (in case of a sat image datacube), a 3D x/y/z voxel cube at a temporal snapshot in case of a climate datacube, or a single value in case of a 1D timeseries.
In short: Axes not touched are not mentioned, and feature subsetting works exactly as before.

**Comment by bradh** [comment]
There is still a burden, in terms of client coders needing to figure it out, implementers needing to include support for it because of test suites, and the general "do I need to understand this or not" barrier to entry. The coders in this case aren't geospatial experts, but rather web developers.

In the `SUBSET` example, that is not how OGC API features implements it - see http://docs.opengeospatial.org/is/17-069r3/17-069r3.html#_parameter_bbox

More generally, arguing that "you only need to implement it if you need it" is basically "Common is just a toolkit", hence my concern about GML.

**Comment by pebau** [comment]
ah, maybe there is a misunderstanding, let me clarify: implementers do not need to support SUBSET in time if they don't want to. WCS, for example, is dimension agnostic. 

Let me cite from http://myogc.org/go/coveragesDWG : "Web developers who have not heard of OGC standards before immediately feel at home with these coverage standards" -- Stephan Siemen, ECMWF

**Comment by cmheazel** [comment]
API-Common already provides a datetime filter.  What more is needed?

**Comment by pebau** [comment]
@cmheazel: tried to explain this above, also with the help of the examples listed.

**Comment by cmheazel** [comment]
@pebau Then perhaps we should bring the full Subsetting conformance class from Coverages into Common. As a separate conformance class, developers don't have to use it if they don't need it.  Yet it would be implemented consistently across APIs which need this functionality.

**Comment by bradh** [comment]
-1.

Where is the "reasonable" limit? If there isn't one defined, we'll end up with Common being a set of conformances classes that is a superset of everything in every API, and back to the GML-like toolkit level complexity. All of those options (call it conformance class, its still variation) add to the cognitive load for everyone, and we're not all as smart as @pebau.

**Comment by cmheazel** [comment]
Added to the Conventions section discussions of the following:
- Coordinate Reference Systems
- Temporal Reference Systems
- Spatial Geometry
- Tempo
Each discussion provides a brief description of the topic, key concepts relevant to API-Common, and references to the governing standards. 
The goal is not re re-invent, but to introduce the concepts and identify their authoritative source.
ral Geometry.

**Comment by cmheazel** [comment]
Updated the discussion in the Conventions section of API-Common:Collections.
Temporal and Spatial CRS discussion has been combined into a single section.
Added a Temporal geometry section.
Grounded these discussions on the "Temporal Thing" and "Spatial Thing" concepts from W3C/OGC Spatial Data on the Web Best Practice)

**Comment by jerstlouis** [comment]
I believe we do need temporal support in Common, perhaps it could be handled through subset query parameter as proposed in #167.

**Comment by cmheazel** [comment]
@jerstlouis Common Part 2 already has the datetime parameter. I agree that a more complex filter should be dealt with in the subset or similar parameter.  Close this issue in deference to #167?

**Comment by jerstlouis** [comment]
@cmheazel One major issue right now is that Part 2 only has a single main conformance class (or the requirements seem organized this way), while in fact it should be multiple, e.g.:

1. Listing available collections
2. Filtering available collections
3. Available query parameters for which data access API would define how they are interpreted to return results

dateTime is defined in https://github.com/opengeospatial/oapi_common/blob/master/collections/requirements/collections/REQ_rc-time-definition.adoc and https://github.com/opengeospatial/oapi_common/blob/master/collections/requirements/collections/REQ_rc-time-response.adoc

But it is not clear to me whether this is 2 or 3 or both. I think date time should be available for both, but that would be separate conformance classes.

e.g. see in Coverages:

https://github.com/opengeospatial/ogc_api_coverages/tree/master/standard/requirements/geodata-filteredList
https://github.com/opengeospatial/ogc_api_coverages/tree/master/standard/requirements/coverage-dateTime

where there are separate requirements for each.

**Comment by ghobona** [comment]
**2020-09-15 OGC API Cross Cutting Issues session during OGC Member Meeting**

The following resolution was agreed upon during the session:

- You, as the implementer, do not have control of which parameters are available in the standard. 
- If each parameter has its  own conformance classes, then the developer can select the parameters to use.
- We should try to make them as course grained as possible. This should be driven by the feedback from developers.

**Comment by cmheazel** [comment]
The current version (5/5/21) defines the parameters n a separate section.  For each operation, it then identifies the applicable parameters and how they are to be applied against the resource associated with that operation.  The result is that the behavior of a parameter is fully defined for each resource where it is applicable.  Derivative standards are free to use the parameters defined in Common, but they must identify the parameter being used, the target resource(s), and the behavior that results from the use of that parameter.  
Note: an operation is a resource path (URI) and the associated HTTP verb.

**Comment by cmheazel** [comment]
5/10/21 SWG - OK to close..
